File: Python/Scripts/ex3_2_1.py
# exercise 3.2.1

from pylab import *
from scipy.io import loadmat
from similarity import similarity

# Image to use as query
i = 888

# Similarity: 'SMC', 'Jaccard', 'ExtendedJaccard', 'Cosine', 'Correlation'
similarity_measure = 'Correlation'

# Load the CBCL face database
# Load Matlab data file to python dict structure
X = loadmat('../Data/wildfaces_grayscale.mat')['X']
N, M = shape(X)


# Search the face database for similar faces
# Index of all other images than i
noti = range(0,i) + range(i+1,N)
# Compute similarity between image i and all others
sim = similarity(X[i,:], X[noti,:], similarity_measure)
sim = sim.tolist()[0]
# Tuples of sorted similarities and their indices
sim_to_index = sorted(zip(sim,noti))


# Visualize query image and 5 most/least similar images
figure(figsize=(12,8))
subplot(3,1,1)

img_hw = int(sqrt(len(X[0])))
imshow(np.reshape(X[i],(img_hw,img_hw)).T, cmap=cm.gray)
xticks([]); yticks([])
title('Query image')
ylabel('image #{0}'.format(i))


for ms in range(5):

    # 5 most similar images found
    subplot(3,5,6+ms)
    im_id = sim_to_index[-ms-1][1]
    im_sim = sim_to_index[-ms-1][0]
    imshow(np.reshape(X[im_id],(img_hw,img_hw)).T, cmap=cm.gray)
    xlabel('sim={0:.3f}'.format(im_sim))
    ylabel('image #{0}'.format(im_id))
    xticks([]); yticks([])
    if ms==2: title('Most similar images')

    # 5 least similar images found
    subplot(3,5,11+ms)
    im_id = sim_to_index[ms][1]
    im_sim = sim_to_index[ms][0]
    imshow(np.reshape(X[im_id],(img_hw,img_hw)).T, cmap=cm.gray)
    xlabel('sim={0:.3f}'.format(im_sim))
    ylabel('image #{0}'.format(im_id))
    xticks([]); yticks([])
    if ms==2: title('Least similar images')

show()

File: Matlab/Scripts/ex3_2_1.m
% exercise 3.2.1

% Image to use as query
i = 1;

% Similarity: 'SMC', 'Jaccard', 'ExtendedJaccard', 'Cosine', 'Correlation' 
SimilarityMeasure = 'cos';

%% Load the CBCL face database
cdir = fileparts(mfilename('fullpath')); 
load(fullfile(cdir,'../Data/wildfaces_grayscale.mat'));
[N,M] = size(X);

%% Search the face database for similar faces

% Index of all other images than i
noti = [1:i-1 i+1:N]; 
% Compute similarity between image i and all others
sim = similarity(X(i,:), X(noti,:), SimilarityMeasure);
% Sort similarities
[val, j] = sort(sim, 'descend');

%% Plot query and result
mfig('Faces: Query'); clf;
subplot(3,5,1:5);
imagesc(reshape(X(i,:),[40,40]));
axis image
set(gca, 'XTick', [], 'YTick', []);
ylabel(sprintf('Image #%d', i));
title('Query image','FontWeight','bold');
for k = 1:5
    subplot(3,5,k+5)
    ii = noti(j(k));
    imagesc(reshape(X(ii,:),[40,40]));
    axis image
    set(gca, 'XTick', [], 'YTick', []);
    xlabel(sprintf('sim=%.2f', val(k)));
    ylabel(sprintf('Image #%d', ii));
    if k==3, title('Most similar images','FontWeight','bold'); end;
end
for k = 1:5
    subplot(3,5,k+10)
    ii = noti(j(end+1-k));
    imagesc(reshape(X(ii,:),[40,40]));
    axis image
    set(gca, 'XTick', [], 'YTick', []);
    xlabel(sprintf('sim=%.3f', val(end+1-k)));
    ylabel(sprintf('Image #%d', ii));
    if k==3, title('Least similar images','FontWeight','bold'); end;
end
colormap(gray);
